### ISSUE TITLE
YOLOv8 with SAHI

### ISSUE THREAD

**Issue by Beckove** [ultralytics/ultralytics#18144]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am currently doing a project for **Small object detection**, the model and algorithm suggested are YOLOv8 and SAHI. But when I train my v8 model with just 150 epochs for a 1700 training images images the model's performance is so insane, it detects mostly all the objects. However, when I apply SAHI on the predicting process, the output seem to be worse (both in number of detections and confidence).
Is SAHI just good in case of we don't have a small enough training dataset and want to detect on an aerial one (like training on farm sheep images and detecting objects in aerial images)
What should I do next?
Thanks for your help!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Beckove, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com) for guidance on using YOLOv8 and various techniques, such as SAHI, where you can find detailed resources for improving small object detection results.

If this is a ❓ Question about custom training or specific use cases like combining YOLOv8 with SAHI, please share additional details to help us assist you better. This could include:

- Examples of your dataset (input images and their annotations)
- Training and prediction logs
- The full code snippet or pipeline you are using to apply SAHI
- Any changes or customizations made to the default settings

For troubleshooting, it's important to ensure you're following best practices for training and prediction. Please review [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to optimize your training and dataset preparation for small object detection.

👉 If you're facing a 🐛 Bug, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) showing the issue, as well as details about the environment and steps taken to apply SAHI. It's important to ensure your software and dependencies are up to date.

## Upgrade

Ensure you're using the latest `ultralytics` package, which may include improvements or fixes related to your issue. Upgrade in a [**Python>=3.8**](https://www.python.org/) environment as shown below:

```bash
pip install -U ultralytics
```

## Environments

YOLOv8 can be run in several verified environments. For a reproducible setup, leverage one of these supported options (with all required software preinstalled):

- **Notebooks** with GPU support: 
  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> 
  <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> 
  <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **AWS** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker** Image: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) 
  <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Community

Join our vibrant community for feedback and support: 
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chat
- [Discourse](https://community.ultralytics.com) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) to connect with fellow users

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing across macOS, Windows, and Ubuntu for all YOLO modes and tasks. 

This is an automated response to get you started. An Ultralytics engineer will follow up with additional assistance soon. 😊

**Comment by Beckove** [comment]
@UltralyticsAssistant, here is my code
`import cv2
from google.colab.patches import cv2_imshow
from sahi.predict import get_sliced_prediction
from sahi.models.yolov8 import Yolov8DetectionModel
import math
image_test=cv2.imread(image_path1)
model=YOLO("/content/gdrive/My Drive/mein_Model/1st_Cow_50/weights/best.pt")
fuck=model.predict(source=image_path1, save=False)
fuck[0].show()
detection_model = Yolov8DetectionModel(
    model_path="/content/gdrive/My Drive/mein_Model/1st_Cow_50/weights/best.pt",                 # Đường dẫn file .pt của bạn
    confidence_threshold=0.1,             # Ngưỡng confidence
    device="cuda" if torch.cuda.is_available() else "cpu"  # GPU nếu khả dụng
)
h,w,c=image_test.shape
# Slicing prediction
result = get_sliced_prediction(
    image=image_test,               # Ảnh đầu vào (numpy array hoặc đường dẫn file)
    detection_model=detection_model, # Mô hình đã được bọc
    slice_height=round(h/1),               # Chiều cao mỗi lát
    slice_width=round(w/1),                # Chiều rộng mỗi lát
    overlap_height_ratio=0.0,       # Tỷ lệ overlap chiều cao
    overlap_width_ratio=0.0         # Tỷ lệ overlap chiều rộng
)
import cv2
import numpy as np
from google.colab.patches import cv2_imshow

# Chuyển đổi ảnh từ PIL sang NumPy array
result_image = np.array(result.image)

# Vẽ bounding boxes lên ảnh
for pred in result.object_prediction_list:
    if pred.score.value > 0.2:  # Lọc theo confidence threshold
        bbox = pred.bbox
        # Vẽ bounding box
        cv2.rectangle(
            result_image,
            (int(bbox.minx), int(bbox.miny)),
            (int(bbox.maxx), int(bbox.maxy)),
            (0, 255, 0),  # Màu xanh lá cây
            2  # Độ dày đường viền
        )
        # Thêm nhãn (nếu cần)
        cv2.putText(
            result_image,
            f"{pred.category.name} {pred.score.value:.2f}",  # Hiển thị nhãn và confidence
            (int(bbox.minx), int(bbox.miny) - 10),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.5,
            (0, 255, 0),
            1
        )

# Hiển thị ảnh đã vẽ bounding boxes
cv2_imshow(result_image)
`
![7c5a66cb-9cae-4a67-bbd1-b5915fe21e13](https://github.com/user-attachments/assets/e3731ad1-53fb-41fc-9ae9-0bf66b095d79)
![8fcea7be-e26b-43f7-8b88-55564a9572ad](https://github.com/user-attachments/assets/a7153040-0c9f-4fc3-afab-5e66c477ff8a)
The blue label is YOLO only, the green one is YOLO+SAHI

**Comment by Y-T-G** [comment]
For training, you should use the imgsz that is equal to the size of the slice that you use while using SAHI.
